core/utils.py

```diff
@@ -1,23 +1,78 @@
 """Utility functions to load data from .mat .yaml and .h5 files."""
 
 import numpy as np
+from jax import numpy as jnp
 import h5py
 import os
 import yaml
 import scipy.io as spio
 import pickle
 from typing import Text
+from pynwb import NWBFile, NWBHDF5IO
+from ndx_pose import PoseEstimationSeries, PoseEstimation
 
+# Constants 
+MM_TO_M = 0.001
 
-def loadmat(filename):
+def daance_to_stac_mjx(data):
     """
-    this function should be called instead of direct spio.loadmat
-    as it cures the problem of not properly recovering python dictionaries
-    from mat files. It calls the function check keys to cure all entries
-    which are still mat-objects
+    Rearrange the data daance .mat mocap ordering
+    to internal order, and scale from millimeters to
+    meters. 
+
+    Args: mocap data arranged in dannce ordering, and
+    stored in mm.
+
+    Out: data rearranged into internal format, scaled to 
+    meters.
+    """
+
+    kp_names = params["KP_NAMES"]
+    # argsort returns the indices that would sort the array
+    stac_keypoint_order = np.argsort(kp_names)
+
+    # Dannce data is stored in mm
+    data = data * MM_TO_M
+    data = jnp.array(data[:, :, stac_keypoint_order])
+    data = jnp.transpose(data, (0, 2, 1))
+    data = jnp.reshape(data, (data.shape[0], -1))
+
+    return data
+
+
+def load_dannce_mat(data_path):
     """
-    data = spio.loadmat(filename, struct_as_record=False, squeeze_me=True)
-    return _check_keys(data)
+    loads in mocap data from .mat file constructed by dannce:
+    (https://github.com/spoonsso/dannce). in particular this means
+    it relies on the data being in millimeters, and that we use the data
+    stored in the "pred" key. 
+    """
+
+    data = spio.loadmat(data_path, struct_as_record=False, squeeze_me=True)
+    data =  _check_keys(data)["pred"][:]
+
+    return daance_to_stac_mjx(data)
+
+
+def load_dannce_nwb(filename):
```

Comment:
```suggestion
def load_nwb(filename):
```

NWB is its own standards-based format, not related to DANNCE.

Reply:
Indeed, though this function is intended to load nwb files created by converting dannce mat files, as in the colab you sent.

Reply:
Ok, though I don't see a specific standard for mocap data. To me this implies that clients are not constrained, and we're possibly open to a Pandora's box of possibilities. Hence, as above, I thought I'd start with the case of dealing with NWB data converted from dannce mat, so at least I could handle our test nwb data. Please let me know if I am wrong.

Reply:
NWB **is a data standard**. That is its entire and singular purpose. From https://www.nwb.org:

> [Neurodata Without Borders](https://www.nwb.org/nwb-neurophysiology/) (NWB) is a **data standard** for neurophysiology, providing neuroscientists with a common standard to share, archive, use, and build analysis tools for neurophysiology data. NWB is designed to store a variety of neurophysiology data, including data from intracellular and extracellular electrophysiology experiments, data from optical physiology experiments, and **tracking** and stimulus data.

---

There are a small number of 3D mocap systems for non-human animals, [ours included](https://github.com/talmolab/sleap-anipose). These are hard to develop and there will not be a huge number of them that spring up. The hope will be that many of them will adapt NWB as an option for export. We should take the initiative to support it as well since it's the only existing data standard specification for 3D mocap data.

DANNCE does not export to NWB, but it should. We converted our test data to NWB manually, but it didn't have to be from DANNCE. There is no logical relationship between NWB and DANNCE formats and we should not couple them.

At the end of the day, 3D mocap data is just a `(frames, nodes, xyz)` shaped array plus some metadata. Our data loaders should just load from different existing formats (starting with the ones we know and use) and get the data into `(frames, nodes, xyz)`. Any other logic, such as reordering, type casting, etc., should be handled downstream.